### PR TITLE
Request action type filter

### PR DESCRIPTION
**Pull request by hellcp-work**
image
Implements a way to filter by action types within the new requests view

Depends on #16536

### CODE REVIEW

File: src/api/app/controllers/webui/request_controller.rb [hunk 1]
```diff
@@ -340,12 +340,18 @@ def set_filter_involvement
...
   end
 
+  def set_filter_action_type
+    @filter_action_type = params[:action_type].presence || []
+    @filter_action_type = @filter_action_type.intersection(BsRequestAction::TYPES)
+  end
+
   def filter_requests
     @bs_requests = filter_by_involvement(@bs_requests, @filter_involvement)
+    @bs_requests = filter_by_action_type(@bs_requests, @filter_action_type)
```
Comment (hennevogel) [inline src/api/app/controllers/webui/request_controller.rb:350]:
```suggestion
    @bs_requests = requests.for_action_type(filter_action_type) if @filter_action_type.present?
```

Comment (hennevogel) [reply src/api/app/controllers/webui/request_controller.rb:350]: No need to hide the scope in some method...

File: src/api/app/controllers/webui/request_controller.rb [hunk 2]
```diff
@@ -340,12 +340,18 @@ def set_filter_involvement
...
   end
 
+  def set_filter_action_type
+    @filter_action_type = params[:action_type].presence || []
+    @filter_action_type = @filter_action_type.intersection(BsRequestAction::TYPES)
+  end
+
   def filter_requests
     @bs_requests = filter_by_involvement(@bs_requests, @filter_involvement)
+    @bs_requests = @bs_requests.for_action_type(@filter_action_type) if @filter_action_type.present?
```
Comment (hennevogel) [inline src/api/app/controllers/webui/request_controller.rb:350]:
```suggestion
    @bs_requests = @bs_requests.bs_request_actions.where(type: @filter_action_type) if @filter_action_type.present?
```

Comment (hennevogel) [reply src/api/app/controllers/webui/request_controller.rb:350]: Or better yet, just make this a scope on the `bs_request_actions` association in `BsRequest`

Comment (hennevogel) [reply src/api/app/controllers/webui/request_controller.rb:350]:
```
has_many :blah do
  def with_types(types: [])
    where(type: types)
  end
end
```

Comment (hennevogel) [reply src/api/app/controllers/webui/request_controller.rb:350]: How is this more complicated? This is how you scope associations the Rails way.

Comment (hennevogel) [reply src/api/app/controllers/webui/request_controller.rb:350]: doing `.where...` is just the "not-shared" way. The scope on the association people will find if they are looking for a way to get actions by type...

Comment (hennevogel) [reply src/api/app/controllers/webui/request_controller.rb:350]: Ah no this is bullshit, we are not filtering the `BsRequestActions` association but `BsRequest`... sorry for the noise.

File: src/api/app/views/webui/request/_request_item.html.haml [hunk 3]
```diff
@@ -1,20 +1,19 @@
 = link_to(request_show_path(bs_request.number), class: 'list-group-item', id: "request-heading#{bs_request.number}") do
   .row.flex-nowrap
     .col-auto
-      %i.fas.fa-code-pull-request
+      - action_icon = action_type_icon(bs_request.bs_request_actions.pluck(:type).group_by(&:itself).values.max_by(&:size).first)
```
Comment (hennevogel) [inline src/api/app/views/webui/request/_request_item.html.haml:4]: Hm, I'm not sure this makes much sense. How about displaying all icons somehow?

**Review by saraycp — APPROVED** [review]

**Comment by saraycp** [comment]
**codecov/patch** complains about the helper, which I don't think it's worth to test, but also about the request controller, in such case we could add a simple spec like the one introduced in #16575.

**Review by saraycp — APPROVED** [review]

**Pull request merged**